### ISSUE TITLE
[CT-27] [Feature] invalidate the packages folder when packages.yml changes

### ISSUE THREAD

**Issue by joellabes** [dbt-labs/dbt-core#4557]
### Is there an existing feature request for this?

- [X] I have searched the existing issues

### Describe the Feature

A frequent issue for both Cloud and Core users is that when you update your packages.yml file, if you don't remember to run `dbt deps` you will stay stuck on the old version. 

In the last couple of days I've seen a couple of cases of this ([1](https://getdbt.slack.com/archives/CBSQTAPLG/p1641535402101800), [2](https://getdbt.slack.com/archives/CMZ2V0X8V/p1641332532034400?thread_ts=1641332093.032600&cid=CMZ2V0X8V)) in the rollout of utils 0.8.0. 

What I'd like to see is something similar to partial parsing being invalidated when dbt_project.yml changes. When you change your packages.yml, if you don't run deps before your next dbt invocation, it should either fail with a clear warning (sort of like what it does when the number of packages listed in the file doesn’t match the number of directories installed), or even classier, run deps for you first. 

### Describe alternatives you've considered

Nothing?

### Who will this benefit?

Pretty much everyone who uses packages has been bitten by this at some time or other.

### Are you interested in contributing this feature?

maybe! Depends on how big a lift it'd be

**Comment by gshank** [comment]
This is a very reasonable request. But running dbt deps on changes is a bigger lift because of the way that our tasks work. We are looking at the possibilities of splitting out the deps code, but that's probably not going to happen in the near future. We could do a similar thing to what we do in partial parsing, the complication there is that it would be easy to put it in the partial parse msgpack file, but that would mean you'd only get the checking when partial parsing is being used, so we'll have to put it in some other sort of state file.

**Comment by joellabes** [comment]
Pragmatic me wants to say it's OK to only do this if partial parsing is enabled, but that sounds like it's signing us up to get a confusing bug report in a few months when someone without partial parsing gets caught out. 

Is there any technical aversion or added complexity to adding another state file for packages?

**Comment by joellabes** [comment]
This just bit me again - I updated my branch on our internal analytics project to get the new `reset_dev_env` macro, and in doing so collected a shiny new `packages.yml` file which added a new Fivetran package. I couldn’t work out why I was getting warnings about not being able to find the referenced model `ad_reporting__url_report` on a brand new clone until I realised that it was provided by a package.

**Comment by jtcohen6** [comment]
@dbeatty and I chatted about this again last week - written-up thoughts forthcoming

**Comment by joellabes** [comment]
I just noticed this log line: 
```
05:22:13  Running with dbt=1.3.0
05:22:13  Unable to do partial parsing because a project dependency has been added
```

does this mean dbt already knows what dependencies exist in a partial parsing context?

**Comment by jtcohen6** [comment]
@joellabes Yes - but only once parsing has begun, and the packages have already been installed. It isn't something that dbt knows as part of the `deps` codepath.

Doug's write-up: https://github.com/dbt-labs/dbt-core/issues/6643

The big idea in "Phase 2" comes down to: If we had a clean record of the packages installed the _last_ time we ran `dbt deps`, we could calculate a diff between the user-configured `packages.yml` and the "resolved" `packages.json`. We could do that using just Hub-requested metadata, without actually downloading any packages. Then, we could do a "slim" `deps` (just update changed packages). Maybe it gets to be so fast that, if we also cached Hub metadata, this could just be part of parsing...

**Comment by iamtodor** [comment]
@joellabes can we please reopen this issue?

**Comment by iamtodor** [comment]
Comment to prevent closing

**Comment by iamtodor** [comment]
Comment to prevent closing

**Comment by graciegoheen** [comment]
dbt-core now generates a `package-lock.yml` file during `dbt deps` and validates installed packages against locked `deps`. See docs [here](https://docs.getdbt.com/reference/commands/deps?version=1.12#predictable-package-installs).

However, users _still_ have to manually run `dbt deps` to detect a change. 

As @joellabes mentioned, we could resolve this a few different ways: 
- If my `packages.yml` file has changes since last invocation of dbt, throw a warning that I need to manually re-run `dbt deps`
- If my `packages.yml` file has changes since last invocation of dbt, automatically run `dbt deps` for me

**Comment by jbbqqf** [comment]
Hi! Triaging older issues this week — I think this one has shipped and can be closed.

**Evidence (verified against current `main`):**

- [`core/dbt/cli/main.py:457`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/cli/main.py#L457) — a new package-lock.yml will be generated and the packages are installed: - user updated the packages.yml
- [`core/dbt/task/deps.py:263`](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/task/deps.py#L263) — this loop is to create the package-lock.yml in the same format as original packages.yml — packages-lock + hash-based invalidation implemented

package-lock.yml functionality regenerates the lock when packages.yml changes, addressing the original 'invalidate packages folder' request. Documented in the deps command help.

Would you mind closing this out? If something specific is still missing — a sub-field, a related behavior — point me at it and I'll dig further.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually against the current `main` branch.)
